File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    min_diff = len(wires)
    
    for [w1, w2] in wires: # 하나씩 끊어보자
        print(f"{w1, w2} 끊음")
        new_wires = [w for w in wires if w != [w1, w2]]
        visited = [False] * (n+1)
        
        q = deque([new_wires[0][0]])
        visited[new_wires[0][0]] = True
        cnt = 1
        
        while q:
            cur = q.popleft()
            print(f"{cur} 방문")
            for nw1, nw2 in new_wires: # 단방향이니까 양방향 확인
                if cur == nw1 and not visited[nw2]:
                    visited[nw2] = True
                    print(nw1, nw2)
                    q.append(nw2)
                    cnt += 1
                elif cur == nw2 and not visited[nw1]:
                    visited[nw1] = True
                    print(nw2, nw1)
                    q.append(nw1)
                    cnt += 1
        
        min_diff = min(min_diff, abs(cnt - (n - cnt)))
        
        print(f"차이: {abs(cnt - (n - cnt))}, 현재 min: {min_diff}")
        
    return min_diff
                

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <queue>
#include <cmath>

using namespace std;

int solution(int n, vector<vector<int>> wires) {
    int ans = n;
    
    for (int i = 0; i < wires.size(); i++) { 
        // 1. 인접 리스트 - i 제외       
        vector<vector<int>> adj(n+1);
        
        for (int j = 0; j < wires.size(); j++) {
            if (i == j) continue;
            int u = wires[j][0];
            int v = wires[j][1];
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        
        // 2. BFS로 네트워크 노드 세기
        vector<bool> visited(n+1, false);
        queue<int> q;
        q.push(1);
        visited[1] = true;
        int cnt = 0;
        
        while(!q.empty()) {
            int cur = q.front();
            q.pop();
            cnt ++;
            
            for(int next : adj[cur]) {
                if (!visited[next]) {
                    visited[next] = true;
                    q.push(next);
                }
            }
        }
        
        int diff = abs(cnt - (n - cnt));
        ans = min(ans, diff);
    }
    
    return ans;
}